R/compound_taskid-utils.R

```diff
@@ -0,0 +1,202 @@
+      round_id = round_id,
+      all_character = TRUE,
+      include_sample_ids = FALSE,
+      bind_model_tasks = FALSE
+    ),
+    function(.x) {
+      if (!has_spls_tbl(.x)) {
+        return(NULL)
+      }
+      dplyr::inner_join(tbl, .x[, names(.x) != out_tid],
+        by = setdiff(names(tbl), out_tid)
+      )
+    }
+  )
+
+  call <- rlang::current_env()
+  tbl_compound_taskids <- purrr::map2(
+    mt_tbls, mt_compound_taskids, function(.x, .y) {
+      get_mt_compound_taskid_set(.x, .y, config_tasks,
+        error = error, call = call
+      )
+    }
+  ) %>%
+    purrr::set_names(seq_along(.))
+
+  if (compact) {
+    tbl_compound_taskids <- purrr::compact(tbl_compound_taskids)
+  }
+
+  tbl_compound_taskids
+}
+
+# Detect the compound_taskid_set for a tbl for a single modeling task.
+get_mt_compound_taskid_set <- function(tbl, config_comp_tids, config_tasks,
+                                       error = TRUE, call = NULL) {
+  if (any(is.null(tbl), is.null(config_comp_tids))) {
+    return(NULL)
+  }
+
+  if (nrow(tbl) == 0L) {
+    return(NULL)
+  }
+  if (is.null(call)) call <- rlang::current_env()
+  out_tid <- hubUtils::std_colnames["output_type_id"]
+  task_ids <- hubUtils::get_task_id_names(config_tasks)
+
+  # Count number of unique values per sample (output_type_id) for each task ID column and
+  # return TRUE if 1 which would indicate a task ID can be considered a compound_taskid
+  spl_uniq <- tbl[, c(task_ids, out_tid)] |>
+    dplyr::group_by(.data[[out_tid]]) |>
+    dplyr::summarise(
+      dplyr::across(
+        dplyr::everything(),
+        function(.x) {
+          dplyr::n_distinct(.x) == 1L
+        }
+      )
+    )
+
+  # Count number of unique values for each task ID column across tbl and
+  # return TRUE if greater than 1.
+  # The aim of this check is to distinguish between false positives, resulting
+  # from task ids that can only take a single value or where only 1 value
+  # is required or all are optional and only one has been provided, so here we also check
+  # whether the value is consistent across other samples. If a single unique value
+  # is consistent across all samples for a task id, it cannot be considered a
+  # compound_taskid and therefore FALSE will be returned. Note that this check
+  # only applies to task ids that are not members of the compound_taskids set.
+  # compound_taskids set vars will always be set to TRUE in this check.
+  tbl_non_uniq <- tbl[, task_ids] |>
+    purrr::map(~ dplyr::n_distinct(.x) > 1L) |>
+    tibble::as_tibble()
+  tbl_non_uniq[, config_comp_tids] <- TRUE
+
+  # Create a sample (output_type_id) vs compound_taskid membership boolean matrix.
+  # For a task id to be considered a compound_taskid, a unique value must be TRUE
+  # at the sample level and we need to ensure it's not a false positive.
+  # Here we combine the two checks.
+  is_compound_taskid <- apply(spl_uniq[, task_ids], 1,
+    function(x) x & tbl_non_uniq[, task_ids],
```

Comment:
For this sample index, for each task id var, TRUE if both:

- there was only one value of that task id var for this sample; and
- the hub named it as a compound task id var OR (the hub didn't name it as a compound task id var AND the model submitted multiple values for it across different samples)

Reply:
this makes sense.